zkml/src/parser/onnx.rs

```diff
@@ -0,0 +1,532 @@
+        ensure_onnx!(dil.iter().all(|&x| x == 1), "Dilations must be 1");
+    }
+    let zkml_maxpool = Layer::Pooling(Pooling::Maxpool2D(Maxpool2D::default()));
+    let node = ProvableNode::new(
+        vec![Edge::new(node.inputs[0].node, node.inputs[0].slot)],
+        zkml_maxpool,
+    );
+    Ok((node_id, node))
+}
+
+fn load_relu(
+    model: &OnnxModel,
+    node_id: NodeId,
+    node: &OnnxNode,
+    _iter: &mut dyn Iterator<Item = &usize>,
+) -> Result<(NodeId, CustomNode), ProvableOpError> {
+    let relu = crate::layers::activation::Relu::new();
+    // find the input node that corresponds to the const input of Relu - since tract_onnx transforms
+    // a relu operation into Max(input, Const(0))
+    // the input node would be the other one.
+    ensure_onnx!(
+        node.inputs.len() == 2,
+        "Relu {} must have 2 inputs",
+        node.name
+    );
+    let real_input_id = match model.node(node.inputs[1].node).op_as::<Const>() {
+        Some(_) => node.inputs[0],
+        None => {
+            ensure_onnx!(
+                model.node(node.inputs[0].node).op_as::<Const>().is_some(),
+                "Relu {} has no constant input",
+                node.name
+            );
+            node.inputs[1]
+        }
+    };
+    let provable_node = crate::layers::provable::ProvableNode::new(
+        vec![Edge::new(real_input_id.node, real_input_id.slot)],
+        Layer::Activation(Activation::Relu(relu)),
+    );
+    Ok((node_id, provable_node))
+}
+
+fn load_gemm(
+    model: &OnnxModel,
+    node_id: NodeId,
+    node: &OnnxNode,
+    iter: &mut dyn Iterator<Item = &usize>,
+) -> Result<(NodeId, CustomNode), ProvableOpError> {
+    let _matrix = downcast_to::<EinSum>(node)?;
+    // TODO: we only support matvec for now for onnx models
+    // Fetch the input which is constant (e.g. the weights)
+    ensure_onnx!(
+        node.inputs.len() == 2,
+        "Gemm {} must have 2 inputs",
+        node.name
+    );
+    let Some(weight_link) = node
+        .inputs
+        .iter()
+        .rev()
+        .find(|&x| is_const(model.node(x.node)))
+    else {
+        return err(format!("Gemm {} has no constant input", node.name));
+    };
+    let weight = extract_const_tensor(model.node(weight_link.node))?;
+    // find the input node
+    let Some(input_link) = node.inputs.iter().find(|&x| x.node != weight_link.node) else {
+        return err(format!("Gemm {} has no input", node.name));
+    };
+
+    // also extract the bias if any. If there is one, that means the next node is a Add node and we
+    // must make sure one of the inputs is the current matrix node. Otherwise that's just a normal add that we don't support.
+    // TODO: support general case with Add layer.
+    let mut piter = iter.peekable();
+    let next_node_id = **piter
+        .peek()
+        .ok_or(ProvableOpError::OnnxParsingError(format!(
+            "Gemm {} has no next node",
+            node.name
+        )))?;
```

Comment:
Couldn't it be the case that Gemm is the last node? Shouldn't we avoid returning an error if there is no next node?

Reply:
Usually a dense layer is always followed by an activation function, but it's true it's not necessary. i'll remove.